From 6ddbe46794b321d4617fc3543922b6f31fd065e3 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Fri, 25 Nov 2022 09:05:45 +0200
Subject: [PATCH] [SW-110961] gaudi2: skip bist if cold rst bit set

This patch adds support to skip sram bist for 48MB sram
if cold rst bit (bist_skip_enable) is set.

FW issues hard reset after running bist. So to avoid rerun
of bist operation preboot sets bist_skip_enable bit in COLD
RESET register. Preboot checks this bit and ignores running
bist if it is set.

Change-Id: I8c3c02dd4ff68e22e174b6b130a38754d5eaab74
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 5031158..12ae783 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -422,11 +422,26 @@
 #define DFT_APB_CONTROL_DEFAULT_VAL		0xFE
 void hl_sram_self_repair(void)
 {
+	struct gaudi2_cold_rst_data cold_rst_data;
+
 	init_sram_bist_data();
+
+	cold_rst_data.data = regmap_readl(mmCOLD_RST_DATA);
 
 	/* Configure SRAM PLLs before running BIST operations */
 	gaudi2_sram_pll_init();
 
+	/* If FW_ALL reset then skip bist */
+	if (hl_rst_src_get() == HL_FW_ALL_RST) {
+		preboot_puts("bist skipped for 48MB SRAM during FW_ALL reset\n");
+		return;
+	}
+
+	if (cold_rst_data.bist_skip_enable) {
+		preboot_puts("bist skipped for 48MB SRAM\n");
+		return;
+	}
+
 	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BIRA);
 	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BISR);
 	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BIST);
